### PR TITLE
Some suggestions to round out the performance advice

### PR DESCRIPTION
**Pull request by ricomariani**
Note this content is very rough.  It seemed to me unlikely that I could meet the style and tone that is desired here and also there might be significant disagreement in principle or priority.

Accordingly, I'm willing to work on this PR to the extent that the editors find it useful but I am also happy to let this PR be abandoned or replaced by something it inspires that is more to your liking.  You may find it easier to harvest the parts you like, rewrite them in your own style, and then apply them as needed.  Perhaps in a different order.

If you would like additional thoughts in one or more areas, feel free to comment and I can add something.  I intend this PR to be a discussion tool.  I cannot foresee it actually being merged anytime soon if ever.

### CODE REVIEW

File: CppCoreGuidelines.md [hunk 1]
```diff
@@ -13791,168 +13793,174 @@ Performance rule summary:
...
 
-##### Note
+##### Reason
 
-Some people optimize out of habit or because it's fun.
+C++ is not the most common choice for situations with no performance demands at
+all, but performance demands are not equal to each other.  Before you can make
+any sensible choices you should understand what success looks like in your own
+domain.  Your requirements likely include some of binary size, latency,
+predictability, power, egress, and perhaps other more unique metrics.  What your
```
Comment (jwakely) [inline CppCoreGuidelines.md:13804]: Something is unique or it isn't, it can't be more unique.

File: CppCoreGuidelines.md [hunk 2]
```diff
@@ -13791,168 +13793,174 @@ Performance rule summary:
...
-
-If there is no need for optimization, the main result of the effort will be more errors and higher maintenance costs.
+### <a name="Rper-reason"></a>Per.1: Understand your requirements and design accordingly.
 
-##### Note
+##### Reason
 
-Some people optimize out of habit or because it's fun.
+C++ is not the most common choice for situations with no performance demands at
+all, but performance demands are not equal to each other.  Before you can make
```
Comment (jwakely) [inline CppCoreGuidelines.md:13801]:
```suggestion
all, but not all performance demands are the same.  Before you can make
```

File: CppCoreGuidelines.md [hunk 3]
```diff
@@ -13791,168 +13793,174 @@ Performance rule summary:
...
+any sensible choices you should understand what success looks like in your own
+domain.  Your requirements likely include some of binary size, latency,
+predictability, power, egress, and perhaps other more unique metrics.  What your
+needs are can and should color the choices you make.  You will likely not be able
+to "optimize" your way to success if your major choices are actually totally unsuitable
+for your goals.  The best practice is to make a sketch of what you indend to do,
+however complex or simple, and then ask yourself "is this likely to meet my needs?"
+If it is, proceed. If it isn't, dig a little more, understand which things will matter
+and invest and or experiment accordingly. When you have a plan that holds water
```
Comment (jwakely) [inline CppCoreGuidelines.md:13810]:
```suggestion
and invest and/or experiment accordingly. When you have a plan that holds water
```

File: CppCoreGuidelines.md [hunk 4]
```diff
@@ -13791,168 +13793,174 @@ Performance rule summary:
...
 
 ##### Reason
 
-Simple code can be very fast. Optimizers sometimes do marvels with simple code
+Time is secondary metric.  Large time values are an effect of either bad
+orchestration, large resource consumption (read too much, use an O(N^2)
+algorithm instead of an O(N) alorigthm), or inefficient use of a resource (i.e.
+asking it to do something it's bad at, such as seeking all over, or dealing with
+non-local memory access).  Typically secondary metrics like time are not
+probative -- "the code is slow, boo hoo" -- primary metrics like "how many
```
Comment (jwakely) [inline CppCoreGuidelines.md:13859]: Probative is not a common word, I would find an alternative way to say this that will be more easily understood.

File: CppCoreGuidelines.md [hunk 5]
```diff
@@ -13791,168 +13793,174 @@ Performance rule summary:
...
+### <a name="Rper-simple"></a>Per.4: Simpler is generally better
 
-##### Example, bad
+Assuming you have done the basics of making sure that your overall design is approximately right,
+such as you are using O(N) if it is available rather than O(N^2) or something like that then you'll
+find that *simplifying* code and data is the best way to make things faster.  This means:
 
-    // intended to be faster, but is often slower
+* don't create complex inheritance graphs
+* don't use complicate dispatch logic if simple logic will do
```
Comment (jwakely) [inline CppCoreGuidelines.md:13895]:
```suggestion
* don't use complicate dispatch logic if simple logic will do
```

File: CppCoreGuidelines.md [hunk 6]
```diff
@@ -13791,168 +13793,174 @@ Performance rule summary:
...
-Low-level code sometimes inhibits optimizations. Optimizers sometimes do marvels with high-level code.
-
-##### Note
-
-???
-
-???
-
-### <a name="Rper-measure"></a>Per.6: Don't make claims about performance without measurements
+The field of performance is littered with myth and bogus folklore.  My favorite aphorism,
```
Comment (jwakely) [inline CppCoreGuidelines.md:13918]: Best to avoid writing in the first person in a document with many authors.

File: CppCoreGuidelines.md [hunk 7]
```diff
@@ -13791,168 +13793,174 @@ Performance rule summary:
...
+"If you're very good at software performance engineering you're only wrong about 95% of the time."
 
-##### Reason
-
-The field of performance is littered with myth and bogus folklore.
 Modern hardware and optimizers defy naive assumptions; even experts are regularly surprised.
 
-##### Note
+Measure. Experiment. Verify.  Accept or disgard in a passionless way.  Repeat as needed.  When needed.
```
Comment (jwakely) [inline CppCoreGuidelines.md:13923]: Discard?

File: CppCoreGuidelines.md [hunk 8]
```diff
@@ -13791,168 +13793,174 @@ Performance rule summary:
...
-With C++20, we can do better still
+We prefer to start with a design that has, shall we say, the right "shape" from the outset.  But this leaves
+many factors to resolve later.  Those small efficiencies Hoare mentioned, 97% of which we can ignore.  The
+3% might kill us.  We don't want to have to redo everything to plug in those few bits that could wait.
 
-    // sortable specifies that c must be a
-    // random-access sequence of elements comparable with <
-    void sort(sortable auto& c);
+These should not change the overall shape of your design.  We really want to start with something that
+is substantially likely to work from the outset.  But you can defer a lot of choices of you make your
```
Comment (jwakely) [inline CppCoreGuidelines.md:13937]:
```suggestion
is substantially likely to work from the outset.  But you can defer a lot of choices if you make your
```

File: CppCoreGuidelines.md [hunk 9]
```diff
@@ -14039,6 +14047,9 @@ a system call,
 a disk access,
 and a message through a network.
 
+Understand the basic costs associated with the most common data types.
+integers, pointers, std::string, std::shared_ptr, std::unique_pointer.  It's hard to make a good design
```
Comment (jwakely) [inline CppCoreGuidelines.md:14051]:
```suggestion
integers, pointers, `std::string`, `std::shared_ptr`, `std::unique_ptr`.  It's hard to make a good design
```

File: CppCoreGuidelines.md [hunk 10]
```diff
@@ -14149,35 +14160,47 @@ This is admittedly rare, but by factoring out a general computation into separat
...
+It doesn't matter which one.
 
 ### <a name="Rper-space"></a>Per.18: Space is time
 
 ##### Reason
 
 Performance is typically dominated by memory access times.
 
-???
+??? seems kind of redunant with the above
```
Comment (jwakely) [inline CppCoreGuidelines.md:14203]:
```suggestion
??? seems kind of redundant with the above
```

**Review by jwakely — CHANGES_REQUESTED** [review]

File: CppCoreGuidelines.md (hunk 6 above)
Comment (ricomariani) [reply CppCoreGuidelines.md:13918]: I'm too used to blogging.  Thank you.

File: CppCoreGuidelines.md (hunk 1 above)
Comment (ricomariani) [reply CppCoreGuidelines.md]: changed unique to exotic.

File: CppCoreGuidelines.md (hunk 5 above)
Comment (ricomariani) [reply CppCoreGuidelines.md]: changed to complicated.

File: CppCoreGuidelines.md (hunk 6 above)
Comment (ricomariani) [reply CppCoreGuidelines.md]: replaced with something 3rd person.  Though I'm quoting myself now... Maybe it's not even worth the aphorism at all.  I would not be too sad if you asked me to delete it entirely.

File: CppCoreGuidelines.md (hunk 7 above)
Comment (ricomariani) [reply CppCoreGuidelines.md]: fixed

File: CppCoreGuidelines.md (hunk 4 above)
Comment (ricomariani) [reply CppCoreGuidelines.md]: reworded

File: CppCoreGuidelines.md (hunk 1 above)
Comment (ricomariani) [reply CppCoreGuidelines.md]: Also thank  you for reading my suggestions and comments.

Comment (jwakely) [reply CppCoreGuidelines.md]: > Everything is unique in the absolute sense but that is only interesting for a conversation over drinks. It isn't good writing.

I'm not going to argue with that :-)

File: CppCoreGuidelines.md (hunk 5 above)
Comment (jwakely) [reply CppCoreGuidelines.md]: Oops, that's what I was trying to suggest but apparently failed to do! I shouldn't do reviews on my phone before breakfast.

**Comment by jwakely** [comment]
Thanks for the micro-fixes. I don't have an opinion whether it should be merged as-is, I'll leave that to the other editors (I'm semi-retired from this project these days). I do think there's a lot of good content here and some of it would be valuable to use.

**Comment by ricomariani** [comment]
I don't even think it should be merged as is. I think it's likely better to cherry pick from the PR and use as they fit.  Really what ever easier is fine with me.  Some of this section is still rough anyway and it's not like it's cast in stone..

**Comment by ricomariani** [comment]
It's fine, the comments were very helpful and this is hardly *advanced* changes we're talking about here😄
________________________________

**Comment by ricomariani** [comment]
After some out-of-band discussion with the editors we decided that it's best for the guidelines to make reference to content like this rather than directly put them in there.  The article this PR inspired https://medium.com/p/f316c9bc4146 is probably a better landing place for this kind of advice.  I recommended that the group harvest some text from here about considering what type of requirements you have and then maybe coloring the guidelines according to those requirements.  i.e. better is relative to your goal so you are likely to want different things if you need binary size vs. throughput vs. power consumption vs. latency.

In any case, it seems appropriate to close this PR now.  It will always be here to harvest.

**Pull request closed without merging**